### PR TITLE
p2p: AddrFetch - don't disconnect on self-announcements

### PR DESCRIPTION
**Pull request by mzumsande**
AddrFetch connections (old name: oneshots) are intended to be short-lived connections on which we ask a peer for addresses via `getaddr` and disconnect after receiving them.

This is done by disconnecting after receiving the first `addr`. However, it is no longer working as intended, because nowadays, the first `addr` a typical bitcoin core node sends is its self-announcement. 
So we'll disconnect before the peer gets a chance to answer our `getaddr`.

I checked that this affects both `-seednode` peers specified manually, and DNS seeds when AddrFetch is used as a fallback if DNS doesn't work for us.

The current behavior of getting peers via AddrFetch when starting with an empty addrman would be to connect to the peer, receive its self-announcement and add it to addrman, disconnect, reconnect to the same peer again as a full outbound (no other addresses in addrman) and then receive more `addr`. This is silly and not in line with AddrFetch peer being intended to be short-lived peers. 

Fix this by only disconnecting after receiving an `addr` message of size > 1.

[Edit] As per review discussion, this PR now also adds a timeout after which we disconnect if we haven't received any suitable `addr`, and a functional test.

### CODE REVIEW

**Comment by sdaftuar** [comment]
I agree this makes sense.  

While you're improving the addrfetch logic, perhaps we could also add a timeout for disconnection so that if an addrfetch peer hasn't responded to our `getaddr` for some amount of time, we eventually disconnect regardless?  I believe right now we'd just stay connected indefinitely to such a peer.

**Review by amitiuttarwar — COMMENTED** [review]
good catch! concept ACK

half baked thought: Wondering if it would make sense to combine this value with the one used [here](https://github.com/bitcoin/bitcoin/blob/123b401e0acf3b26f149711bdf06d7f6eced1968/src/net_processing.cpp#L2772). In my understanding, both are trying to use the number of addrs to discern [response to getaddr] vs [probably a self-announcement]. We could introduce a constant that's used in both places.

I'm also wondering if we can add functional test coverage for `AddrFetch` connections. I think the groundwork is in place with the outbound-full-relay and block-relay-only connections. We might just be able to add the addr-fetch option to `TestNode.add_outbound_p2p_connection`, the `addconnection` RPC, and `CConnman::AddConnection`. But I only took an initial look & this certainly does not need to happen in this PR.

**Comment by sipa** [comment]
Concept ACK

**Review by lsilva01 — COMMENTED** [review]
Tested ACK https://github.com/bitcoin/bitcoin/pull/22096/commits/b6c5d1e450dde6a54bd785504c923adfb45c7060. 
I ran the node with the args `-debug=net -dnsseed=0 -seednode=<IP> -fixedseeds=0` using the master branch and this PR branch (table below).

The way `ADDR_FETCH` is used today to fetch new address from a seed IP seems not work as expected, since the first message is only self-announcement and the `ADDR` message with 1000 addresses is received via `OUTBOUND_FULL_RELAY` from the same peer.

As shown in the second table, after the proposed change in this PR, the first `ADDR_FETCH` connection actually fetches for addresses. 

Master Branch results:
| From IP  | Connection Type |  `vAddr` size | `AddrMan` size |  self-announcement
| ------------- | ------------- | ------------- | ------------- | ------------- |
| `seed_ip`  | `ADDR_FETCH`  | 1 | 0 | yes
| `seed_ip`   | `OUTBOUND_FULL_RELAY`  | 1 | 1 | yes
| `seed_ip`   | `OUTBOUND_FULL_RELAY`  | 1000 | 1 | no
| `new_ip`   | `OUTBOUND_FULL_RELAY`  | 1 | 870 | yes

PR Branch results:
| From IP  | Connection Type |  `vAddr` size | `AddrMan` size |  self-announcement
| ------------- | ------------- | ------------- | ------------- | ------------- |
| `seed_ip` | `ADDR_FETCH`  | 1 | 0 | yes
| `seed_ip` | `ADDR_FETCH`  | 1000 | 1 | no
| `new_ip`   | `OUTBOUND_FULL_RELAY`  | 1 | 870 | yes

**Comment by Gondayt** [comment]
I gondayt never agreed to let anyone else speak for me or sell my bitcoins? I have 11 withdraws off me and the 11th one is to git hubi will pull it and show y'all maybe you can lete know what's going on

**Comment by Gondayt** [comment]
> Tested ACK [b6c5d1e](https://github.com/bitcoin/bitcoin/commit/b6c5d1e450dde6a54bd785504c923adfb45c7060).
> I ran the node with the args `-debug=net -dnsseed=0 -seednode=<IP> -fixedseeds=0` using the master branch and this PR branch (table below).
> 
> The way `ADDR_FETCH` is used today to fetch new address from a seed IP seems not work as expected, since the first message is only self-announcement and the `ADDR` message with 1000 addresses is received via `OUTBOUND_FULL_RELAY` from the same peer.
> 
> As shown in the second table, after the proposed change in this PR, the first `ADDR_FETCH` connection actually fetches for addresses.
> 
> Master Branch results:
> 
> From IP	Connection Type	`vAddr` size	`AddrMan` size	self-announcement
> `seed_ip`	`ADDR_FETCH`	1	0	yes
> `seed_ip`	`OUTBOUND_FULL_RELAY`	1	1	yes
> `seed_ip`	`OUTBOUND_FULL_RELAY`	1000	1	no
> `new_ip`	`OUTBOUND_FULL_RELAY`	1	870	yes
> PR Branch results:
> 
> From IP	Connection Type	`vAddr` size	`AddrMan` size	self-announcement
> `seed_ip`	`ADDR_FETCH`	1	0	yes
> `seed_ip`	`ADDR_FETCH`	1000	1	no
> `new_ip`	`OUTBOUND_FULL_RELAY`	1	870	yes

**Comment by mzumsande** [comment]
> While you're improving the addrfetch logic, perhaps we could also add a timeout for disconnection so that if an addrfetch peer hasn't responded to our `getaddr` for some amount of time, we eventually disconnect regardless? I believe right now we'd just stay connected indefinitely to such a peer.

Yes, I think that would be useful, and I will add this. I think such an addrfetch peer also wouldn't count toward the max full outbound limit in spite of effectively behaving like one. As for the timeout, I'd suggest a value `4*AVG_ADDRESS_BROADCAST_INTERVAL = 2min` after connection.

> half baked thought: Wondering if it would make sense to combine this value with the one used here. In my understanding, both are trying to use the number of addrs to discern [response to getaddr] vs [probably a self-announcement]. We could introduce a constant that's used in both places.

Do you suggest just using `size > 10` instead of 1 in the check? More seems difficult, because the rest of the criteria work on the level of individual addrs.

> I'm also wondering if we can add functional test coverage for `AddrFetch` connections. I think the groundwork is in place with the outbound-full-relay and block-relay-only connections. We might just be able to add the addr-fetch option to TestNode.add_outbound_p2p_connection, the addconnection RPC, and CConnman::AddConnection
timeout behavior is tested. 

Good idea! Adding AddrFetch to the test framework actually does seem to work exactly as you suggested, just need to figure out some details and will add this in the next days (unless reviewers prefer that to be a separate PR).

**Comment by mzumsande** [comment]
I added the timeout and a functional test now. Thanks a lot @amitiuttarwar for help with the functional test!

File: src/net_processing.cpp [hunk 1]
```diff
@@ -4364,6 +4366,12 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
 
     MaybeSendPing(*pto, *peer, current_time);
 
+    if(pto->IsAddrFetchConn() && current_time - std::chrono::seconds(pto->nTimeConnected) > 4 * AVG_ADDRESS_BROADCAST_INTERVAL) {
```
Comment (jonatack) [inline src/net_processing.cpp:4369]: nit, don't hesitate to run clang format on the changes
```suggestion
    if (pto->IsAddrFetchConn() && current_time - std::chrono::seconds(pto->nTimeConnected) > 4 * AVG_ADDRESS_BROADCAST_INTERVAL) {
```

File: src/rpc/net.cpp [hunk 2]
```diff
@@ -337,7 +337,7 @@ static RPCHelpMan addconnection()
         "\nOpen an outbound connection to a specified node. This RPC is for testing only.\n",
         {
             {"address", RPCArg::Type::STR, RPCArg::Optional::NO, "The IP address and port to attempt connecting to."},
-            {"connection_type", RPCArg::Type::STR, RPCArg::Optional::NO, "Type of connection to open, either \"outbound-full-relay\" or \"block-relay-only\"."},
+            {"connection_type", RPCArg::Type::STR, RPCArg::Optional::NO, "Type of connection to open, \"outbound-full-relay\", \"block-relay-only\" or \"addr-fetch\"."},
```
Comment (jonatack) [inline src/rpc/net.cpp:340]:
```suggestion
            {"connection_type", RPCArg::Type::STR, RPCArg::Optional::NO,
             "Type of connection to open (\"outbound-full-relay\", \"block-relay-only\" or \"addr-fetch\")."},
```

File: test/functional/p2p_addrfetch.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ADDR.nServices = NODE_NETWORK | NODE_WITNESS
+ADDR.ip = "192.0.0.8"
+ADDR.port = 18444
+
+
+class AddrFetchPeer(P2PInterface):
+    def __init__(self):
+        super().__init__()
+        self.got_getheaders = False
+        self.got_getaddr = False
```
Comment (jonatack) [inline test/functional/p2p_addrfetch.py:27]: New variables naming nit, `s/got_*/received_*/`

File: test/functional/p2p_addrfetch.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    def run_test(self):
+        node = self.nodes[0]
+        self.log.info('Connect to an addr-fetch peer')
+        peer = node.add_outbound_p2p_connection(AddrFetchPeer(), p2p_idx=0, connection_type="addr-fetch")
+        info = node.getpeerinfo()
+        assert_equal(len(info), 1)
+        assert_equal(info[0]['connection_type'], 'addr-fetch')
+
+        self.log.info('Check that we send getaddr but don\'t try to sync headers with the addr-fetch peer')
```
Comment (jonatack) [inline test/functional/p2p_addrfetch.py:47]:
```suggestion
        self.log.info("Check that we send getaddr but don't try to sync headers with the addr-fetch peer")
```

File: test/functional/p2p_addrfetch.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        peer.sync_send_with_ping
```
Comment (jonatack) [inline test/functional/p2p_addrfetch.py:48]:
```suggestion
        peer.sync_send_with_ping()
```

File: test/functional/p2p_addrfetch.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+        self.log.info('Check that answering the getaddr with a single address does not lead to disconnect')
+        # This prevents disconnecting on self-announcements
+        msg = msg_addr()
+        msg.addrs = [ADDR]
+        peer.send_and_ping(msg)
+        assert_equal(len(node.getpeerinfo()), 1)
+
+        self.log.info('Check that answering with larger addr messages leads to disconnect')
+        msg.addrs = [ADDR]*2
```
Comment (jonatack) [inline test/functional/p2p_addrfetch.py:60]: PEP8
```suggestion
        msg.addrs = [ADDR] * 2
```

File: test/functional/p2p_addrfetch.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        peer.send_message(msg)
+        peer.wait_for_disconnect(timeout=5)
+
+        self.log.info('Check timeout for addr-fetch peer that do not send addrs')
```
Comment (jonatack) [inline test/functional/p2p_addrfetch.py:64]:
```suggestion
        self.log.info('Check timeout for addr-fetch peer that does not send addrs')
```

File: test/functional/p2p_addrfetch.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        peer = node.add_outbound_p2p_connection(AddrFetchPeer(), p2p_idx=1, connection_type="addr-fetch")
+        node.setmocktime(int(time.time()) + 130)  # Timeout: 2 min
+        peer.sync_send_with_ping
```
Comment (jonatack) [inline test/functional/p2p_addrfetch.py:67]:
```suggestion
        peer.sync_send_with_ping()
```

File: test/functional/test_framework/test_node.py [hunk 9]
```diff
@@ -557,9 +557,9 @@ def add_p2p_connection(self, p2p_conn, *, wait_for_verack=True, **kwargs):
         return p2p_conn
 
     def add_outbound_p2p_connection(self, p2p_conn, *, p2p_idx, connection_type="outbound-full-relay", **kwargs):
-        """Add an outbound p2p connection from node. Either
-        full-relay("outbound-full-relay") or
-        block-relay-only("block-relay-only") connection.
+        """Add an outbound p2p connection from node. Must be a
+        full-relay("outbound-full-relay"), block-relay-only("block-relay-only")
```
Comment (jonatack) [inline test/functional/test_framework/test_node.py:561]:
```suggestion
        full-relay ("outbound-full-relay"), block-relay-only ("block-relay-only")
```
or: `Must be an "outbound-full-relay", "block-relay-only" or "addr-fetch" connection.`

File: src/net.cpp [hunk 10]
```diff
@@ -1166,16 +1166,27 @@ void CConnman::CreateNodeFromAcceptedSocket(SOCKET hSocket,
...
+        case ConnectionType::FEELER:
+            return false;
+        case ConnectionType::OUTBOUND_FULL_RELAY:
+            max_connections = m_max_outbound_full_relay;
+        case ConnectionType::BLOCK_RELAY:
+            max_connections = m_max_outbound_block_relay;
+        // no limit for ADDR_FETCH because -seednode has no limit either
+        case ConnectionType:: ADDR_FETCH:
+            max_connections = 0;
+    }
```
Comment (jonatack) [inline src/net.cpp:1182]: perhaps give `max_connections` a value on initialization and add a comment per doc/developer-notes.md
```diff
 bool CConnman::AddConnection(const std::string& address, ConnectionType conn_type)
 {
-    int max_connections;
+    int max_connections{0};
     switch(conn_type) {
         case ConnectionType::INBOUND:
         case ConnectionType::MANUAL:
@@ -1178,8 +1178,8 @@ bool CConnman::AddConnection(const std::string& address, ConnectionType conn_typ
             max_connections = m_max_outbound_block_relay;
         // no limit for ADDR_FETCH because -seednode has no limit either
         case ConnectionType:: ADDR_FETCH:
-            max_connections = 0;
-    }
+            break;
+    } // no default case, so the compiler can warn about missing cases
```

**Review by jonatack — COMMENTED** [review]
Approach ACK d728c6c8717960d4cbc5c2ea10833ce997d7ae1b some optional suggestions below, nice find and thanks for adding the test coverage.

File: src/net_processing.cpp (hunk 1 above)
Comment (mzumsande) [reply src/net_processing.cpp]: done, thanks

File: src/rpc/net.cpp (hunk 2 above)
Comment (mzumsande) [reply src/rpc/net.cpp]: done

File: test/functional/p2p_addrfetch.py (hunk 5 above)
Comment (mzumsande) [reply test/functional/p2p_addrfetch.py]: oops, thanks!

File: test/functional/p2p_addrfetch.py (hunk 7 above)
Comment (mzumsande) [reply test/functional/p2p_addrfetch.py]: fixed

File: test/functional/p2p_addrfetch.py (hunk 8 above)
Comment (mzumsande) [reply test/functional/p2p_addrfetch.py]: Changed to 121 (more leeway is not really needed) and removed the `sync_send_with_ping()` which is incorrect here because it conflicts with the disconnect.

File: test/functional/test_framework/test_node.py (hunk 9 above)
Comment (mzumsande) [reply test/functional/test_framework/test_node.py]: Took the short version, that should be clear enough.

File: src/net.cpp (hunk 10 above)
Comment (mzumsande) [reply src/net.cpp]: done

**Comment by mzumsande** [comment]
Thanks for the review @jonatack! I took your suggestions and pushed an update.

**Comment by sipa** [comment]
utACK 535c990e16c6c191aa1b484b800a16bf5e4d0765

File: test/functional/p2p_addrfetch.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class AddrFetchPeer(P2PInterface):
+    def __init__(self):
+        super().__init__()
+        self.received_getheaders = False
+        self.received_getaddr = False
+
+    def on_getheaders(self, message): self.received_getheaders = True
+    def on_getaddr(self, message): self.received_getaddr = True
```
Comment (jnewbery) [inline test/functional/p2p_addrfetch.py:30]: No need for this specialized subclass. The base P2PInterface counts messages received by type, so you can just use that:

```diff
diff --git a/test/functional/p2p_addrfetch.py b/test/functional/p2p_addrfetch.py
index b50a710adb..e50e9e7ee4 100755
--- a/test/functional/p2p_addrfetch.py
+++ b/test/functional/p2p_addrfetch.py
@@ -9,7 +9,7 @@ Test p2p addr-fetch connections
 import time
 
 from test_framework.messages import msg_addr, CAddress, NODE_NETWORK, NODE_WITNESS
-from test_framework.p2p import P2PInterface
+from test_framework.p2p import P2PInterface, p2p_lock
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import assert_equal
 
@@ -19,17 +19,6 @@ ADDR.nServices = NODE_NETWORK | NODE_WITNESS
 ADDR.ip = "192.0.0.8"
 ADDR.port = 18444
 
-
-class AddrFetchPeer(P2PInterface):
-    def __init__(self):
-        super().__init__()
-        self.received_getheaders = False
-        self.received_getaddr = False
-
-    def on_getheaders(self, message): self.received_getheaders = True
-    def on_getaddr(self, message): self.received_getaddr = True
-
-
 class P2PAddrFetch(BitcoinTestFramework):
 
     def set_test_params(self):
@@ -39,15 +28,16 @@ class P2PAddrFetch(BitcoinTestFramework):
     def run_test(self):
         node = self.nodes[0]
         self.log.info("Connect to an addr-fetch peer")
-        peer = node.add_outbound_p2p_connection(AddrFetchPeer(), p2p_idx=0, connection_type="addr-fetch")
+        peer = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=0, connection_type="addr-fetch")
         info = node.getpeerinfo()
         assert_equal(len(info), 1)
         assert_equal(info[0]['connection_type'], 'addr-fetch')
 
         self.log.info("Check that we send getaddr but don't try to sync headers with the addr-fetch peer")
         peer.sync_send_with_ping()
-        assert peer.received_getaddr
-        assert not peer.received_getheaders
+        with p2p_lock:
+            assert peer.message_count['getaddr'] >= 1
+            assert peer.message_count['getheaders'] == 0
 
         self.log.info("Check that answering the getaddr with a single address does not lead to disconnect")
         # This prevents disconnecting on self-announcements
@@ -62,7 +52,7 @@ class P2PAddrFetch(BitcoinTestFramework):
         peer.wait_for_disconnect(timeout=5)
 
         self.log.info("Check timeout for addr-fetch peer that does not send addrs")
-        peer = node.add_outbound_p2p_connection(AddrFetchPeer(), p2p_idx=1, connection_type="addr-fetch")
+        peer = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=1, connection_type="addr-fetch")
         node.setmocktime(int(time.time()) + 121)  # Timeout: 2 minutes
         peer.wait_for_disconnect(timeout=5)
```

(this also fixes a potential race where data in the `P2PInterface` was being read without locking the p2p_lock)

**Review by jnewbery — COMMENTED** [review]
Code review ACK 535c990e16c6c191aa1b484b800a16bf5e4d0765

One suggestion to simplify the test

File: test/functional/p2p_addrfetch.py (hunk 11 above)
Comment (mzumsande) [reply test/functional/p2p_addrfetch.py]: Thanks - I simplified the test as suggested. (with the slightly stricter `assert peer.message_count['getaddr'] == 1`)

**Comment by jnewbery** [comment]
Code review ACK 8dbf2b210dbdaddc11295b926fefddca127d7e8a

**Review by amitiuttarwar — COMMENTED** [review]
code review ACK 8dbf2b210d

**Comment by mzumsande** [comment]
Analyzing #22243 gave me concern the concern that a timeout of `4*AVG_ADDRESS_BROADCAST_INTERVAL = 2min` may be  too low, given the distribution of `PoissonNextSend()`, leading to 1) the node disconnecting early before the peer would send an addr too often 2) an intermittently failing test. I'll do some simulations later today to confirm.

**Comment by mzumsande** [comment]
I changed the timeout to `10 * AVG_ADDRESS_BROADCAST_INTERVAL = 5min`. Simulated this, and while the probability of failure (i.e. disconnecting before the peer has scheduled its getaddr answer) was ~1.8% before, it is now of order 10^-5 which I think is fine.

**Comment by practicalswift** [comment]
Concept ACK

Good catch!

**Comment by amitiuttarwar** [comment]
reACK 96061d3410898d4759488e50d1f78211849ede50

**Comment by naumenkogs** [comment]
Concept ACK

**Comment by jnewbery** [comment]
Code review ACK 96061d3410898d4759488e50d1f78211849ede50

File: test/functional/p2p_addrfetch.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        # This prevents disconnecting on self-announcements
+        msg = msg_addr()
+        msg.addrs = [ADDR]
+        peer.send_and_ping(msg)
+        assert_equal(len(node.getpeerinfo()), 1)
+
+        self.log.info("Check that answering with larger addr messages leads to disconnect")
+        msg.addrs = [ADDR] * 2
+        peer.send_message(msg)
+        peer.wait_for_disconnect(timeout=5)
```
Comment (lsilva01) [inline test/functional/p2p_addrfetch.py:53]: Is the `assert_equal()` necessary for the validation?

```suggestion
        peer.wait_for_disconnect(timeout=5)
        assert_equal(len(node.getpeerinfo()), 0)
```

File: test/functional/p2p_addrfetch.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+        self.log.info("Check timeout for addr-fetch peer that does not send addrs")
+        peer = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=1, connection_type="addr-fetch")
+        node.setmocktime(int(time.time()) + 301)  # Timeout: 5 minutes
+        peer.wait_for_disconnect(timeout=5)
```
Comment (lsilva01) [inline test/functional/p2p_addrfetch.py:58]: Is the `assert_equal()` necessary for the validation?

```suggestion
        peer.wait_for_disconnect(timeout=5)
        assert_equal(len(node.getpeerinfo()), 0)
```

**Review by lsilva01 — APPROVED** [review]
re-ACK https://github.com/bitcoin/bitcoin/pull/22096/commits/96061d3410898d4759488e50d1f78211849ede50

File: test/functional/p2p_addrfetch.py (hunk 12 above)
Comment (amitiuttarwar) [reply test/functional/p2p_addrfetch.py:53]: checking the node's connections doesn't seem harmful, but it also doesn't seem necessary because of the way the `wait_for_disconnect` function works- it keeps checking if the P2PConnection object [is connected](https://github.com/bitcoin/bitcoin/blob/master/test/functional/test_framework/p2p.py#L154), and would fail if that was still true after 5 seconds.

Comment (mzumsande) [reply test/functional/p2p_addrfetch.py:53]: I agree - will remove the line here and below if I touch again.

Comment (mzumsande) [reply test/functional/p2p_addrfetch.py:53]: I agree that `wait_for_disconnect` should be sufficient.

File: test/functional/p2p_addrfetch.py (hunk 13 above)
Comment (mzumsande) [reply test/functional/p2p_addrfetch.py:58]: same as [here](https://github.com/bitcoin/bitcoin/pull/22096#discussion_r655799680), `wait_for_disconnect` should be sufficient.

File: src/net_processing.cpp [hunk 14]
```diff
@@ -4364,6 +4364,12 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
 
     const auto current_time = GetTime<std::chrono::microseconds>();
 
+    if (pto->IsAddrFetchConn() && current_time - std::chrono::seconds(pto->nTimeConnected) > 10 * AVG_ADDRESS_BROADCAST_INTERVAL) {
```
Comment (naumenkogs) [inline src/net_processing.cpp:4367]: Why linking this timing to `AVG_ADDRESS_BROADCAST_INTERVAL`? It seems irrelevant.
Also, 5 minutes feels like a lot, if we thinking of a bootstrapping node hitting buggy/sybil nodes again and again.

Comment (naumenkogs) [reply src/net_processing.cpp:4367]: Ah yeah, I see the Poisson justification. Shouldn't it be linked to `AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL` then? This is the relevant constant for the first Poisson delay, no?

File: src/net_processing.cpp [hunk 15]
```diff
@@ -2779,7 +2779,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
         m_addrman.Add(vAddrOk, pfrom.addr, 2 * 60 * 60);
         if (vAddr.size() < 1000) peer->m_getaddr_sent = false;
```
Comment (naumenkogs) [inline src/net_processing.cpp:2781]: Does this work as intended though? 
1. Alice connects to Bob, each other send GETADDR, m_getaddr_sent=true
2. Alice self-announces, Bob sets `m_getaddr_sent=false`
3. Alice responds to GETADDR, `m_getaddr_sent` is already false, so the check relying on it is not really correct.

The check being incorrect seems to be not critical because `vAddr.size() <= 10` is very likely to be false anyway (assuming sane addrman), but what's the point then?

What this check actually does is `don't relay the first self-announcement`. I don't think this is intended? Or if it is, it's very unclear from the codebase esp with that variable name. We could really benefit from a comment.

File: src/net.cpp [hunk 16]
```diff
@@ -1166,16 +1166,29 @@ void CConnman::CreateNodeFromAcceptedSocket(SOCKET hSocket,
...
+        break;
+    } // no default case, so the compiler can warn about missing cases
 
     // Count existing connections
     int existing_connections = WITH_LOCK(cs_vNodes,
                                          return std::count_if(vNodes.begin(), vNodes.end(), [conn_type](CNode* node) { return node->m_conn_type == conn_type; }););
 
     // Max connections of specified type already exist
-    if (existing_connections >= max_connections) return false;
+    if (max_connections && existing_connections >= max_connections) return false;
```
Comment (naumenkogs) [inline src/net.cpp:1191]: nit: I would probably slightly std::optional over relying on `bool(0) == false`

File: src/net_processing.cpp (hunk 15 above)
Comment (naumenkogs) [reply src/net_processing.cpp:2781]: Note: what i'm asking here is not really related to the change of this PR.

File: src/net_processing.cpp (hunk 14 above)
Comment (mzumsande) [reply src/net_processing.cpp:4367]: I don't think so: `AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL=24h` only deals with our self-announcements, which are then added to `m_addrs_to_send`. The frequency with which we actually send out the content of `m_addrs_to_send` is controlled by  `AVG_ADDRESS_BROADCAST_INTERVAL`. 

I initially chose a timeout of 2min instead of 5min, but that would have lad to a premature disconnect in 1.8% of attempts, which seemed too high.
I think it makes sense to be a bit more conservative here, because we use AddrFetch only when we have specified the peer with `-seednode` or when we can't do DNS and thus connect manually to DNS seeds. We currently don't make AddrFetch connections to random nodes, and don't just choose another AddrFetch peer if we don't get an answer from the current one.

File: src/net_processing.cpp (hunk 15 above)
Comment (mzumsande) [reply src/net_processing.cpp:2781]: I agree, it doesn't work as intended! There was some earlier discussion about this (the flag was recently renamed from `fGetAddr`), see https://github.com/bitcoin/bitcoin/pull/21707#issuecomment-824361280 and you may recall #19794. Maybe it would be good to revisit this topic...

File: src/net.cpp (hunk 16 above)
Comment (mzumsande) [reply src/net.cpp]: Done.

File: src/net_processing.cpp (hunk 14 above)
Comment (naumenkogs) [reply src/net_processing.cpp:4367]: ACK the second part justification.

Re first part: You are right, my confusion was because of 2 variables we have: `m_next_addr_send` and `m_next_local_addr_send`.

Comment (naumenkogs) [inline src/net_processing.cpp:4367]: So I'm just pointing out here that this timeout also handles the case when they send us 1-addr messages (in a malicious case, this can be anything, not just self-announcement).

Currently I can't come up with a way why this is bad, but if we want to disable this unintended behaviour, we could disconnect addrfetch peers on the second 1-addr message too.

File: src/net.cpp [hunk 17]
```diff
@@ -1166,16 +1166,29 @@ void CConnman::CreateNodeFromAcceptedSocket(SOCKET hSocket,
...
+        break;
+    } // no default case, so the compiler can warn about missing cases
 
     // Count existing connections
     int existing_connections = WITH_LOCK(cs_vNodes,
                                          return std::count_if(vNodes.begin(), vNodes.end(), [conn_type](CNode* node) { return node->m_conn_type == conn_type; }););
 
     // Max connections of specified type already exist
-    if (existing_connections >= max_connections) return false;
+    if (max_connections != std::nullopt && existing_connections >= max_connections) return false;
```
Comment (jnewbery) [inline src/net.cpp:1191]: nit: it may be slightly more idiomatic to use `std::optional::has_value()`, or even just use the [bool operator](https://en.cppreference.com/w/cpp/utility/optional/operator_bool):

```suggestion
    if (max_connections && existing_connections >= max_connections) return false;
```

**Review by jnewbery — COMMENTED** [review]
ACK 5730a43703

File: src/net_processing.cpp (hunk 14 above)
Comment (amitiuttarwar) [reply src/net_processing.cpp:4367]: @naumenkogs do I understand correctly that you're concerned about the case where an AddrFetch connection could repeatedly send addr messages with 1 address in them, until this timeout is hit? 

The malicious case seems very unlikely- we only make AddrFetch connections to nodes submitted via `-seednode`, or to the DNS seeds in certain circumstances. 

In a more ordinary case, I could imagine how the gossiping of self-announcements could lead to > 1 addr message with a size of 1. If the AddrFetch connection runs a Bitcoin Core node, it seems possible since this line waits approximately 10 intervals before timing out. But also I believe the seeders often run custom software and it doesn't make sense to me that they should be less effective if they were to participate in network gossip. 

Conceptually, AddrFetch connections are trying to get a `getaddr` to be fulfilled, and a `getaddr` is a request for many addresses. To me, it makes sense for the requesting node to wait a certain amount of time for an acceptable response, and disconnect if that condition is not met. I'm not seeing why we would want to disconnect even earlier in this circumstance.

**Comment by amitiuttarwar** [comment]
reACK 5730a43703f7e5a5ca26245ba3b55fbdd027d0b6

File: src/net_processing.cpp (hunk 14 above)
Comment (naumenkogs) [reply src/net_processing.cpp:4367]: >The malicious case seems very unlikely- we only make AddrFetch connections to nodes submitted via -seednode, or to the DNS seeds in certain circumstances.

Yes indeed. For some reason I thought we do this to random nodes periodically. Maybe that's a feature I wanted to see, in some form :)
I agree that for current master it's much less of a concern. A comment might help for future when master changes.

> If the AddrFetch connection runs a Bitcoin Core node, it seems possible since this line waits approximately 10 intervals before timing out.

I don't understand this part. Yeah it's 10 intervals of 30s, but self-announcement happens once every 24h. Very-very unlikely a second self-announcement appears within 300s.

>But also I believe the seeders often run custom software and it doesn't make sense to me that they should be less effective if they were to participate in network gossip.

I also don't understand this part, what does it have to do with them participating in gossip?

Comment (amitiuttarwar) [reply src/net_processing.cpp:4367]: I was entertaining the idea of when a node might send multiple addr messages of size 1 before fulfilling the getaddr message. Although bitcoin core nodes will only initiate a self announcement once every ~24 hours, they will relay self announcements from other nodes.

Comment (mzumsande) [reply src/net_processing.cpp:4367]: I think it's also not really more malicious than when they send us nothing at all: The worst thing I could think of that they might do is to painstakingly send us more than 1000 addresses this way.

I'm not sure what kind of comment would help, I'd rather not describe the AddrFetch success-case logic in detail here at the failure case, because if we adjust that logic, the comment will most likely be forgotten.  
Maybe something non-specific like "Timeout for AddrFetch peers for when they don't sent us a suitable getaddr response (in which case we'd immediately disconnect)"?

File: src/net.cpp (hunk 17 above)
Comment (mzumsande) [reply src/net.cpp:1191]: Will change to `std::optional::has_value()` when I push next time - I don't like the bool operator for `optional<int>` because it being true in the case `max_connections == 0` might be a bit confusing for a cursory reader.

**Comment by naumenkogs** [comment]
ACK 5730a43703f7e5a5ca26245ba3b55fbdd027d0b6

**Comment by laanwj** [comment]
Posthumous ACK 5730a43703f7e5a5ca26245ba3b55fbdd027d0b6

File: src/net.cpp (hunk 17 above)
Comment (MarcoFalke) [inline src/net.cpp:1191]: I find it slightly confusing to check for `std::nullopt`, but then use the >= operator on std::optional. That operator will again check for `std::nullopt`. Funnily that operator will evaluate to true if `max_connections` was `std::nullopt`, which it can't be here. It might be more clear to just use the >= operator on `int`.

Comment (mzumsande) [reply src/net.cpp:1191]: I see your point - do I understand it correctly that you suggest to use `existing_connections >= max_connections.value()` in the comparison instead? I could do that as a follow-up.

Comment (MarcoFalke) [reply src/net.cpp:1191]: `*max_connections` should be enough because an exception can't be thrown

File: test/functional/p2p_addrfetch.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        assert_equal(len(node.getpeerinfo()), 1)
+
+        self.log.info("Check that answering with larger addr messages leads to disconnect")
+        msg.addrs = [ADDR] * 2
+        peer.send_message(msg)
+        peer.wait_for_disconnect(timeout=5)
+
+        self.log.info("Check timeout for addr-fetch peer that does not send addrs")
+        peer = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=1, connection_type="addr-fetch")
+        node.setmocktime(int(time.time()) + 301)  # Timeout: 5 minutes
```
Comment (jonatack) [inline test/functional/p2p_addrfetch.py:57]: The state just before the timeout could be checked to ensure the disconnection only occurs after the 5 minute threshold.

**Review by jonatack — COMMENTED** [review]
Post-humous ACK 5730a43703f7e5a5ca26245ba3b55fbdd027d0b6
